File: stacy_frontend/lib/src/models/plant_data.dart
class PlantData {
  // final int healthPercentage;
  final DateTime timestamp;
  final double temperature;
  final double humidity;
  final double moisture;
  final double pressure;
  final double hic;
  final double batteryVoltage;
  final double batteryPercentage;

  PlantData({
    required this.timestamp,
    required this.temperature,
    required this.humidity,
    required this.moisture,
    required this.pressure,
    required this.hic,
    required this.batteryVoltage,
    required this.batteryPercentage,
  });

  factory PlantData.fromJson(Map<String, dynamic> json) {
    if (json.isEmpty) {
      throw ArgumentError('JSON data for the PlantData cannot be empty');
    }
    for (var key in [
      'temperature',
      'humidity',
      'moisture',
      'pressure',
      'hic',
      'batteryVoltage',
      'batteryPercentage',
      'timestamp'
    ]) {
      if (!json.containsKey(key)) {
        throw ArgumentError('JSON data for PlantData is missing "$key" key');
      }
    }

    if (json['temperature'] == null || json['temperature'] is! num) {
      throw ArgumentError('Invalid or missing "temperature" in JSON data');
    }
    if (json['humidity'] == null || json['humidity'] is! num) {
      throw ArgumentError('Invalid or missing "humidity" in JSON data');
    }
    if (json['moisture'] == null || json['moisture'] is! num) {
      throw ArgumentError('Invalid or missing "moisture" in JSON data');
    }
    if (json['pressure'] == null || json['pressure'] is! num) {
      throw ArgumentError('Invalid or missing "pressure" in JSON data');
    }
    if (json['hic'] == null || json['hic'] is! num) {
      throw ArgumentError('Invalid or missing "hic" in JSON data');
    }
    if (json['batteryVoltage'] == null || json['batteryVoltage'] is! num) {
      throw ArgumentError('Invalid or missing "batteryVoltage" in JSON data');
    }
    if (json['batteryPercentage'] == null ||
        json['batteryPercentage'] is! num) {
      throw ArgumentError(
          'Invalid or missing "batteryPercentage" in JSON data');
    }
    if (json['timestamp'] == null || json['timestamp'] is! String) {
      throw ArgumentError('Invalid or missing "timestamp" in JSON data');
    }

    return PlantData(
      timestamp: DateTime.parse(json['timestamp'] as String),
      temperature: (json['temperature'] as num).toDouble(),
      humidity: (json['humidity'] as num).toDouble(),
      moisture: (json['moisture'] as num).toDouble(),
      pressure: (json['pressure'] as num).toDouble(),
      hic: (json['hic'] as num).toDouble(),
      batteryVoltage: (json['batteryVoltage'] as num).toDouble(),
      batteryPercentage: (json['batteryPercentage'] as num).toDouble(),
    );
  }
}

File: backend/models/PlantData.js
/**
 * Represents a weather data reading from a client device.
 */
class PlantData {
  /**
   *
   * Creates an instance of PlantData.
   * @param {number} temperature - The temperature reading in Celsius.
   * @param {number} humidity - The humidity reading in percentage.
   * @param {number} moisture - The soil moisture reading in percentage.
   * @param {number} pressure - The atmospheric pressure reading in hPa.
   * @param {number} hic - The heat index in Celsius.
   * @param {number} batteryVoltage - The battery voltage in volts.
   * @param {number} batteryPercentage - The battery percentage.
   */
  constructor(
    temperature,
    humidity,
    moisture,
    pressure,
    hic,
    batteryVoltage,
    batteryPercentage
  ) {
    if (typeof temperature !== 'number' || isNaN(temperature)) {
      throw new Error('Invalid or missing temperature: must be a number.');
    }
    if (typeof humidity !== 'number' || isNaN(humidity)) {
      throw new Error('Invalid or missing humidity: must be a number.');
    }
    if (typeof moisture !== 'number' || isNaN(moisture)) {
      throw new Error('Invalid or missing moisture: must be a number.');
    }
    if (typeof pressure !== 'number' || isNaN(pressure)) {
      throw new Error('Invalid or missing pressure: must be a number.');
    }
    if (typeof hic !== 'number' || isNaN(hic)) {
      throw new Error('Invalid or missing hic: must be a number.');
    }
    if (typeof batteryVoltage !== 'number' || isNaN(batteryVoltage)) {
      throw new Error('Invalid or missing batteryVoltage: must be a number.');
    }
    if (typeof batteryPercentage !== 'number' || isNaN(batteryPercentage)) {
      throw new Error(
        'Invalid or missing batteryPercentage: must be a number.'
      );
    }

    this.temperature = temperature;
    this.humidity = humidity;
    this.moisture = moisture;
    this.pressure = pressure;
    this.hic = hic;
    this.batteryVoltage = batteryVoltage;
    this.batteryPercentage = batteryPercentage;
  }

  /**
   * Returns a plain object representation of the WeatherData instance.
   * This is useful for sending as JSON or inserting into databases.
   * @returns {object} Plain object with weather data.
   */
  toObject() {
    const obj = {
      temperature: this.temperature,
      humidity: this.humidity,
      moisture: this.moisture,
      pressure: this.pressure,
      hic: this.hic,
      batteryVoltage: this.batteryVoltage,
      batteryPercentage: this.batteryPercentage,
    };
    return obj;
  }

  /**
   * Static factory method to create a PlantData instance from a raw object.
   * @param {object} rawData - The raw data object, typically from req.body.
   * @param {number} rawData.temperature
   * @param {number} rawData.humidity
   * @param {number} rawData.moisture
   * @param {number} rawData.pressure
   * @param {number} rawData.hic
   * @param {number} rawData.batteryVoltage
   * @param {number} rawData.batteryPercentage
   * @returns {PlantData}
   * @throws {Error} if validation fails.
   */
  static fromObject(rawData) {
    if (!rawData) {
      throw new Error('Raw data object is required.');
    }
    return new PlantData(
      rawData.temperature,
      rawData.humidity,
      rawData.moisture,
      rawData.pressure,
      rawData.hic,
      rawData.batteryVoltage,
      rawData.batteryPercentage
    );
  }
}

module.exports = PlantData;
